crates/bevy_math/src/bezier.rs

```diff
@@ -0,0 +1,376 @@
+                // Using Newton's method, use the tangent line to estimate a better guess value.
+                let slope = self.dx_dt(t_guess);
+                t_guess -= error / slope;
+                false
+            }
+        });
+        t_guess.clamp(0.0, 1.0)
+    }
+}
+
+/// Generic implementations for sampling Bezier curves. Consider using the methods on [`Bezier`] for
+/// more ergonomic use.
+pub mod generic {
+    use super::Point;
+
+    /// Compute the Bernstein basis polynomial for iteration `i`, for a Bezier curve with with
+    /// degree `n`, at `t`.
+    #[inline]
+    pub fn bernstein_basis(n: usize, i: usize, t: f32) -> f32 {
+        (1. - t).powi((n - i) as i32) * t.powi(i as i32)
+    }
+
+    /// Efficiently compute the binomial coefficient of `n` choose `k`.
+    #[inline]
+    fn binomial_coeff(n: usize, k: usize) -> usize {
+        let k = usize::min(k, n - k);
+        (0..k).fold(1, |val, i| val * (n - i) / (i + 1))
+    }
+
+    /// Evaluate the Bezier curve B(t) of degree `N-1` at the parametric value `t`.
+    #[inline]
+    pub fn position<P: Point, const N: usize>(control_points: [P; N], t: f32) -> P {
+        let p = control_points;
+        let degree = N - 1;
+        (0..=degree)
+            .map(|i| p[i] * binomial_coeff(degree, i) as f32 * bernstein_basis(degree, i, t))
+            .sum()
+    }
+
+    /// Compute the first derivative B'(t) of Bezier curve B(t) of degree `N-1` at the given
+    /// parametric value `t` with respect to t. Note that the first derivative of a Bezier is also a
+    /// Bezier, of degree `N-2`.
+    #[inline]
+    pub fn velocity<P: Point, const N: usize>(control_points: [P; N], t: f32) -> P {
+        if N <= 1 {
+            return P::default(); // Zero for numeric types
+        }
+
+        let p = control_points;
+        let degree = N - 1;
+        let degree_vel = N - 2; // the velocity Bezier is one degree lower than the position Bezier
+        (0..=degree_vel)
+            .map(|i| {
+                // Point on the velocity Bezier curve:
+                let p = (p[i + 1] - p[i]) * degree as f32;
+                p * binomial_coeff(degree_vel, i) as f32 * bernstein_basis(degree_vel, i, t)
+            })
+            .sum()
+    }
+
+    /// Compute the second derivative B''(t) of Bezier curve B(t) of degree `N-1` at the given
+    /// parametric value `t` with respect to t. Note that the second derivative of a Bezier is also
+    /// a Bezier, of degree `N-3`.
+    #[inline]
+    pub fn acceleration<P: Point, const N: usize>(control_points: [P; N], t: f32) -> P {
+        if N <= 2 {
+            return P::default(); // Zero for numeric types
+        }
+        let p = control_points;
+        let degree = N - 1;
+        let degree_vel = N - 2; // the velocity Bezier is one degree lower than the position Bezier
+        let degree_accel = N - 3; // the accel Bezier is one degree lower than the velocity Bezier
+        (0..degree_vel)
```

Comment:
Curious if the compiler will unroll this properly and elide bounds checking or not. Not worth blocking on but definitely something worth looking into.

Reply:
This isn't substitute for looking at the compiled ASM, but when I was benchmarking and optimizing, I found that these iterators were faster than equivalent for loops. I assumed it was for this reason.

Reply:
Benchmarking building a cubic Bezier with 100 points, where those points are acceleration or position:

```
build_pos_cubic_100_points
                        time:   [384.77 ns 385.17 ns 385.60 ns]
Found 15 outliers among 100 measurements (15.00%)
  8 (8.00%) low severe
  2 (2.00%) low mild
  1 (1.00%) high mild
  4 (4.00%) high severe

build_accel_cubic_100_points
                        time:   [384.53 ns 386.51 ns 388.01 ns]
Found 3 outliers among 100 measurements (3.00%)
  1 (1.00%) low severe
  2 (2.00%) low mild
```